### PR TITLE
Alerting: Classic conditions can now display multiple values

### PR DESCRIPTION
**Pull request by gotjosh**
Related #46095

Most of the context it's in the issue above, but the tl;dr is that we can support extracting values from classic conditions.

Classic conditions do have what we call `refID` but unlike expressions, all conditions are expressed under the same `refID`. Each condition can have very different values depending on how we want the values to be reduced (`last`, `avg`, etc).

In this PR we're extending the capture of values to also support classic conditions.

**Special notes for your reviewer**:

- [x] update documentation

### CODE REVIEW

File: pkg/services/ngalert/eval/extract_md.go [hunk 1]
```diff
@@ -66,17 +68,35 @@ func extractEvalString(frame *data.Frame) (s string) {
 	return ""
 }
 
-// extractValues returns the RefID and value for all reduce and math expressions
-// in the frame. It does not return values for classic conditions as the values
-// in classic conditions do not have a RefID. It returns nil if there are
-// no results in the frame.
+// extractValues returns the RefID and value for all classic conditions, reduce, and math expressions in the frame
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/eval/extract_md.go:71]:
```suggestion
// extractValues returns the RefID and value for all classic conditions, reduce, and math expressions in the frame.
```

File: pkg/services/ngalert/state/state.go [hunk 2]
```diff
@@ -33,8 +33,8 @@ type Evaluation struct {
 	EvaluationTime  time.Time
 	EvaluationState eval.State
 	// Values contains the RefID and value of reduce and math expressions.
-	// It does not contain values for classic conditions as the values
-	// in classic conditions do not have a RefID.
+	// Classic conditions can have different values for the same RefID as they include multiple conditions.
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/state/state.go:36]:
```suggestion
	// Classic conditions can have different values for the same RefID as they can include multiple conditions.
```

File: pkg/services/ngalert/state/state.go [hunk 3]
```diff
@@ -33,8 +33,8 @@ type Evaluation struct {
...
+	// For those, we use the index of the condition in addition RefID as the key e.g. "A0, A1, A2, etc.".
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/state/state.go:37]:
```suggestion
	// For these, we use the index of the condition in addition RefID as the key e.g. "A0, A1, A2, etc.".
```

**Review by grobinson-grafana — APPROVED** [review]

File: pkg/services/ngalert/eval/extract_md.go [hunk 4]
```diff
@@ -66,17 +68,35 @@ func extractEvalString(frame *data.Frame) (s string) {
 	return ""
 }
 
-// extractValues returns the RefID and value for all reduce and math expressions
-// in the frame. It does not return values for classic conditions as the values
-// in classic conditions do not have a RefID. It returns nil if there are
-// no results in the frame.
+// extractValues returns the RefID and value for all classic conditions, reduce, and math expressions in the frame.
+// For classic conditions the same refID can have multiple values due to multiple conditions, for them we use the index of
```
Comment (achatterjee-grafana) [inline pkg/services/ngalert/eval/extract_md.go:72]:
```suggestion
// For classic conditions the same refID can have multiple values for multiple conditions. We use the index of
```

File: pkg/services/ngalert/eval/extract_md.go [hunk 5]
```diff
@@ -66,17 +68,35 @@ func extractEvalString(frame *data.Frame) (s string) {
...
+// the condition in addition to the refID to distinguish between them.
```
Comment (achatterjee-grafana) [inline pkg/services/ngalert/eval/extract_md.go:73]:
```suggestion
// the condition in addition to the refID to distinguish between different values.
```

File: pkg/services/ngalert/eval/extract_md.go [hunk 6]
```diff
@@ -66,17 +68,35 @@ func extractEvalString(frame *data.Frame) (s string) {
...
+// It returns nil if there are no results in the frame.
```
Comment (achatterjee-grafana) [inline pkg/services/ngalert/eval/extract_md.go:74]:
```suggestion
// It will return nil when there are no results in the frame.
```

File: docs/sources/alerting/unified-alerting/alerting-rules/alert-annotation-label.md [hunk 7]
```diff
@@ -31,8 +31,8 @@ Labels are key-value pairs that contain information about, and are used to uniqu
...
+| ------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| $labels | The labels from the query or condition. For example, `{{ $labels.instance }}` and `{{ $labels.job }}`. This is unavailable when the rule uses a [classic condition]({{< relref "./create-grafana-managed-rule/#single-and-multi-dimensional-rule" >}}).                                                                                                                                        |
+| $values | The values of all reduce and math expressions that were evaluated for this alert rule. For example, `{{ $values.A }}`, `{{ $values.A.Labels }}` and `{{ $values.A.Value }}` where `A` is the `refID` of the expression. If the rule uses classic conditions, then a combination of the `refID` and position of the condition is used e.g. `{{ $values.A0.Value }}` or `{{ $values.A1.Value }}` |
```
Comment (achatterjee-grafana) [inline docs/sources/alerting/unified-alerting/alerting-rules/alert-annotation-label.md:37]:
```suggestion
| $values | The values of all reduce and math expressions that were evaluated for this alert rule. For example, `{{ $values.A }}`, `{{ $values.A.Labels }}` and `{{ $values.A.Value }}` where `A` is the `refID` of the expression. If the rule uses classic conditions, then a combination of the `refID` and position of the condition is used. For example, `{{ $values.A0.Value }}` or `{{ $values.A1.Value }}` |
```

**Review by achatterjee-grafana — COMMENTED** [review]
Added doc review.

**Review by achatterjee-grafana — APPROVED** [review]

**Review by yuri-tceretian — APPROVED** [review]

**Comment by gotjosh** [comment]
Spoke to @yuri-tceretian offline about this and he had some really good questions about whether we could improve the reference and don't rely on position.

tl;dr is that it is not possible without a) a massive rework of how we retrieve/store operations (which includes the front-end) and b) keeping a similar structure to how we do for expressions.

As an example, this is how we store operations on the database.

```json
[
  {
    "refId": "A",
    "queryType": "",
    "relativeTimeRange": {
      "from": 600,
      "to": 0
    },
    "datasourceUid": "wL7iH1pnk",
    "model": {
      "exemplar": false,
      "expr": "vector(10) + time() % 50",
      "hide": false,
      "interval": "",
      "intervalMs": 1000,
      "legendFormat": "",
      "maxDataPoints": 43200,
      "refId": "A"
    }
  },
  {
    "refId": "B",
    "queryType": "",
    "relativeTimeRange": {
      "from": 0,
      "to": 0
    },
    "datasourceUid": "-100",
    "model": {
      "conditions": [
        {
          "evaluator": {
            "params": [
              0
            ],
            "type": "gt"
          },
          "operator": {
            "type": "and"
          },
          "query": {
            "params": [
              "A"
            ]
          },
          "reducer": {
            "params": [],
            "type": "last"
          },
          "type": "query"
        },
        {
          "evaluator": {
            "params": [
              1,
              0
            ],
            "type": "gt"
          },
          "operator": {
            "type": "and"
          },
          "query": {
            "params": [
              "A"
            ]
          },
          "reducer": {
            "params": [],
            "type": "median"
          },
          "type": "query"
        }
      ],
      "datasource": {
        "type": "__expr__",
        "uid": "-100"
      },
      "hide": false,
      "intervalMs": 1000,
      "maxDataPoints": 43200,
      "refId": "B",
      "type": "classic_conditions"
    }
  }
]
```

You can see that using something like the operation + refID (e.g. `avg(A)`) as an identifier wouldn't work as the conditions can have the same operation (e.g. both could be `avg`). If there are better suggestions, I'm all ears.
